### ISSUE TITLE
[feature request] Check if a file is successfully sent

### ISSUE THREAD

**Issue by bhavya32** [VISWESWARAN1998/Simple-Yet-Hackable-WhatsApp-api#56]
@VISWESWARAN1998 I  want to add a very important feature request.(for me it is important.) Documents can be big or small. If the document is too big, the selenium will close before the file is uploaded. If we increase the time, the next file may be small and it will waste time. We should together sort this issue. There must be some workaround to check if the message is sent.

**Comment by aahnik** [comment]
# I recommend you try sending documents in a different yet SIMPLE manner

- **Upload the documents to _Google Drive_**
- **Get a Link** 
(Only the people to  whom you will give the  link can view) 
 **put the link in the message**

And it's done 😊

**Comment by aahnik** [comment]
**WhatsApp is great for notifying and updating people** 

To send documents the best way is to send emails 😊 with attachments.

# check out AutoMailer

##  [GitHub Link ](GitHub.com/aahnik/AutoMailer)

1. **Send dynamic emails with unlimited variables pulling data from a database(csv file)**

2. **Supports Markdown Formatting & embed links or images**

3. **Supports Attaching any kind of files**

**Comment by bhavya32** [comment]
@aahnik alternatives are not a solution for me. I am mostly requested by my customers to send PDF of invoices and ledgers. Some of them have not saved my number. Whatsapp flags it and they can not click on the link. In short, The link is clickable only if the other party has saved my number. For email, I do send mails but only 5% of them do have mail.

**Comment by aahnik** [comment]
I understand your situation

I am working on it. I have figured out a solution.

Wait @bhavya32 , your problem will be solved

**Comment by aahnik** [comment]
I have made changes solving #56 and #54 part 7 solved #48 in a modular approach for send_message, send_document,send_picture
# @bhavya32  i tried to solve your problem see my pull request #53 added new commits recently 
## do a test run 
you can see my version of whatsapp api which i forked from @VISWESWARAN1998 [here](https://github.com/aahnik/Simple-Yet-Hackable-WhatsApp-api)

**Comment by aahnik** [comment]
##  @bhavya32  i tried to solve

>  I want to add a very important feature request.(for me it is important.) Documents can be big or small. If the document is too big, the selenium will close before the file is uploaded. If we increase the time, the next file may be small and it will waste time. We should together sort this issue. There must be some workaround to check if the message is sent.

In pull  request #53 added new commits recently 

_**Using webdriver to wait until presence of element located so timing is dynamic **_

**Comment by VISWESWARAN1998** [comment]
@aahnik @bhavya32 

I have tested with the 50MB document and I cannot reproduce the issue with my existing implementation as I can able to send files without problem. 

I would like to know more on this so that I can able to fix it.

Thanks and Regards,
Visweswaran N

**Comment by aahnik** [comment]
@bhavya32  you said that selenium will close before sending file

Check your internet speed.
Try override timeout

@VISWESWARAN1998 your internet and system may be faster overall , so you did not face the issue #56 by @bhavya32 .
Even I did not face the issue.

But the issue may occur if file loads slowly in someone's setup.

So webdriver must wait until presence of element located.
Refer to my last few commits in #53

**Comment by aahnik** [comment]
I never faced the issue #56 so I can't really check whether my new commits are actually solving it.

It works for me with both old and new code ...

**Comment by aahnik** [comment]
> @aahnik alternatives are not a solution for me. I am mostly requested by my customers to send PDF of invoices and ledgers. Some of them have not saved my number. Whatsapp flags it and they can not click on the link. In short, The link is clickable only if the other party has saved my number. For email, I do send mails but only 5% of them do have mail. 

Hi @bhavya32 ,

Even if your customers have not saved your number , you can send links ( I have tested ) 
Yes they can open links
Tips for not being flagged by whatsapp.

1.Save your customers phone numbers in your phone 

( ☝️Feel free to Contact me for automation)


2.  Explicitly Tell your customers to save your number for smooth delivery of messages

3. Upload files to gdrive and share those links.
(You may use Google drive or sheets or docs api s for automation)

4. If you run a large business contact Facebook , register and Verify your company. Then you can use Docker based WhatsApp official API , to make chatbots or send messages.( Not totally free but best for large business)

5. Use WhatsApp automatic responder apps in Play Store

These ways might help 

If you provide a description of your processes and systems I can help better.

**Comment by aahnik** [comment]
Hi @bhavya32 , you can try using KhataBook App for accounting.

I don't know whether it's suitable for your setup or not.

But KhataBook can send free SMS and WhatsApp reminders and invoices

**Comment by bhavya32** [comment]
@VISWESWARAN1998 No no! I am not facing the issue directly. I forgot to mention that I kill the chromedriver as well as chrome after the message is sent. As soon as selenium returns, and the file is uploading, I have put a timer of say 5 seconds. It is essential as if I rerun the script without chromedriver, the second chrome instance does not work. @aahnik thanks for suggestions but I do not think any of them would help.

**Comment by aahnik** [comment]
# why do you kill chromedriver everytime..??
> if I rerun the script without closing chrome, the second chrome instance does not work
use the code correctly , u wont need to kill chromedriver

```python
from whatsapp import WhatsApp
whatsapp = WhatsApp(10) # do not repeat this 

while (condition):
        # send message
        # send document
```

**Comment by bhavya32** [comment]
Thanks a lot @aahnik and @VISWESWARAN1998 for looking into my issue. I very well know each and every alternatives and do not need any. As for why I kill chrome-driver, I have implemented it as an API. For example - "127.0.0.1/default?phn=1234567890&message=ABC" would automatically open and send messages. Such an API can be integrated in most applications easily and I was using it to send automatic account updates to people. I only required to know if python could check if "Tick-Mark" appeared in the last message some-how. I do not think it is possible so I am closing the issue.
